timemachine/md/moves.py

```diff
@@ -12,69 +14,96 @@
 from timemachine.lib import custom_ops
-from timemachine.md.barostat.utils import get_bond_list, get_group_indices
 from timemachine.md.states import CoordsVelBox
-from timemachine.potentials import BoundPotential, HarmonicBond
+from timemachine.potentials import BoundPotential
 
 _State = TypeVar("_State")
 
 
-class MonteCarloMove(Generic[_State]):
-    n_proposed: int = 0
-    n_accepted: int = 0
+class Move(Generic[_State], ABC):
+    @abstractmethod
+    def move(self, _: _State) -> _State:
+        ...
+
+    def sample_chain_iter(self, x: _State) -> Iterator[_State]:
+        while True:
+            x = self.move(x)
+            yield x
+
+    def sample_chain(self, x: _State, n_samples: int) -> List[_State]:
+        return list(islice(self.sample_chain_iter(x), n_samples))
+
 
+@dataclass
+class MonteCarloMove(Move[_State], ABC):
+
+    _n_proposed: int = field(default=0, init=False)
+    _n_accepted: int = field(default=0, init=False)
+
+    @abstractmethod
     def propose(self, x: _State) -> Tuple[_State, float]:
         """return proposed state and log acceptance probability"""
-        raise NotImplementedError
 
     def move(self, x: _State) -> _State:
         proposal, log_acceptance_probability = self.propose(x)
-        self.n_proposed += 1
+        self._n_proposed += 1
 
         alpha = np.random.rand()
         acceptance_probability = np.exp(log_acceptance_probability)
         if alpha < acceptance_probability:
-            self.n_accepted += 1
+            self._n_accepted += 1
             return proposal
         else:
             return x
 
     @property
-    def acceptance_fraction(self):
-        if self.n_proposed > 0:
-            return self.n_accepted / self.n_proposed
-        else:
-            return 0.0
+    def n_proposed(self) -> int:
+        return self._n_proposed
+
+    @property
+    def n_accepted(self) -> int:
+        return self._n_accepted
+
+    @property
+    def acceptance_fraction(self) -> float:
+        return self._n_accepted / self._n_proposed if self._n_proposed else np.nan
 
 
-class CompoundMove(MonteCarloMove[_State]):
-    def __init__(self, moves: List[MonteCarloMove]):
-        """Apply each of a list of moves in sequence"""
-        self.moves = moves
+class MetropolisHastingsMove(MonteCarloMove[_State], ABC):
+    @abstractmethod
+    def propose_with_log_q_diff(self, x: _State) -> Tuple[_State, float]:
+        """Return proposed state and the difference in log unnormalized probability, i.e.
+
+        log_q_diff = log(q(x_proposed)) - log(q(x))
+        """
+
+    def propose(self, x: _State) -> Tuple[_State, float]:
+        proposal, log_q_diff = self.propose_with_log_q_diff(x)
+        log_acceptance_probability = np.minimum(log_q_diff, 0.0)
+        return proposal, log_acceptance_probability
+
+
+@dataclass
+class CompoundMove(Move[_State]):
+    """Apply each of a list of moves in sequence"""
+
+    moves: Sequence[MonteCarloMove[_State]]
```

Comment:
Should this be `Sequence[Move[_State]]` else the doc string should clarify that it is only `MonteCarloMove`s

Reply:
The annotation is as intended, since `n_accepted_by_move` etc. only make sense for `MonteCarloMove`s. Updated the docstring in https://github.com/proteneer/timemachine/pull/1134/commits/d634b45c8e93ecfb33159dbb2d29120e480904e0 to hopefully clarify.